File: java/src/main/java/org/katas/checkout/controller/CheckoutController.java
package org.katas.checkout.controller;

import org.katas.checkout.controller.dtos.CheckoutResponseDto;
import org.katas.checkout.controller.dtos.CheckoutRequestDto;
import org.katas.checkout.controller.exceptions.BadRequestException;
import org.katas.checkout.services.CheckoutService;
import org.katas.checkout.services.model.CheckoutRequest;

public class CheckoutController {

    private final CheckoutService checkoutService;

    CheckoutController(CheckoutService checkoutService) {
        this.checkoutService = checkoutService;
    }

    public CheckoutResponseDto checkout(CheckoutRequestDto requestDto) {
        if (requestDto == null || requestDto.cartId() == null)
            throw new BadRequestException();

        var response = checkoutService.checkout(mapToCheckoutRequest(requestDto));
        return new CheckoutResponseDto(response.orderId(), null, -1);
    }

    private CheckoutRequest mapToCheckoutRequest(CheckoutRequestDto requestDto) {
        return new CheckoutRequest(requestDto.cartId(), requestDto.deliveryDto().name(), requestDto.deliveryDto().street(), requestDto.deliveryDto().city(), requestDto.creditCardDto().number(), requestDto.creditCardDto().expirationDate(), requestDto.creditCardDto().ccv());
    }
}

File: ts/src/09-outside-in/api/checkout.controller.ts
import {CheckoutRequestDto} from "./dtos/checkout-request.dto";
import {CheckoutResponseDto} from "./dtos/checkout-response.dto";
import {BadRequestException} from "./dtos/bad-request-exception";
import {ICheckoutService} from "../services/checkout-service.interface";
import {CheckoutServiceParams} from "../model/checkoutServiceParams";

export class CheckoutController {
    constructor(private checkoutService: ICheckoutService) {
    }


    checkout(request: CheckoutRequestDto): CheckoutResponseDto {
        if (!request.basketId) {
            throw new BadRequestException('basketId is required');
        }
        const response = this.checkoutService.doCheckout(new CheckoutServiceParams(
            request.basketId
        ));
        return {orderId: response.orderId} as CheckoutResponseDto;
    }
}
